assets/js/base/hocs/with-query-string-values.js

```diff
@@ -0,0 +1,92 @@
+import { Component } from 'react';
+import { addQueryArgs, getQueryArg } from '@wordpress/url';
+
+/**
+ * HOC that keeps the state in sync with the URL query string.
+ */
+const withQueryStringValues = ( values ) => ( OriginalComponent ) => {
+	let instances = 0;
+
+	class WrappedComponent extends Component {
+		urlParameterSuffix = instances++ > 0 ? `_${ instances }` : '';
```

Comment:
It's not immediately clear to me what the purpose of this instances tracking is.  I see from the [tests for this block](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1040/files#diff-fdbab8ee1c29bcab9e4bc6b1efab2d9dR28) **what** it does but I'm having trouble grasping **why**.  I think it may be the tests throwing me off a bit.

Reply:
I added a comment explaining why this is required (795ca60d0caf78946a9afa62e89c7eaf6d4db080) and updated the test description (783d3b6454d94995aa25808d481a1d2534125724). Let me know if it makes sense.

Reply:
Hmm okay, so essentially the query string might just be `?foo=bar` but if multiple components are reading that query string, the hoc will append instance vars to the param key for each component to work with?  What problem does this solve though.  Why can't components read from the same query string?  Sorry, but it's still not clear to me what problems are being solved here.  

The comments are a bit more descriptive in describing what is happening but I'm still unclear on the why.

Reply:
Basically this allows having two or more _All Products_ blocks in the same page and each one behave independently of the others.

Otherwise, all of them would write and read from the same query string values causing unexpected results.